File: benchmarks/weird_captcha_gym/shared_runtime/server/incubator_graders/popup_exorcist.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "popup_exorcist"


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID or str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id or str(public_state.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "stale challenge"}
    events = payload.get("events")
    if not isinstance(events, list) or not 5 <= len(events) <= 180:
        return {"graded": True, "passed": False, "feedback": "containment transcript is missing or outside limits"}
    originals = set(str(item) for item in ground_truth.get("popup_ids") or [])
    parasite = str(ground_truth.get("parasite_id") or "")
    echoes = set(str(item) for item in ground_truth.get("echo_ids") or [])
    popup_by_id = {str(item.get("id")): dict(item) for item in public_state.get("popups") or []}
    live, infected = set(originals), set()
    provoked = False
    spawn_pending = False
    contained = ""
    last_drag: dict[str, list[list[int]]] = {}
    purged = False
    for sequence, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return {"graded": True, "passed": False, "feedback": f"event {sequence} sequence mismatch"}
        kind, window_id = str(event.get("kind") or ""), str(event.get("window_id") or "")
        if spawn_pending and kind != "spawn":
            return {"graded": True, "passed": False, "feedback": "parasite echoes were not recorded immediately"}
        if kind == "focus":
            if window_id not in live:
                return {"graded": True, "passed": False, "feedback": "focused window was not live"}
            continue
        if kind == "close":
            if window_id not in live:
                return {"graded": True, "passed": False, "feedback": "closed window was not live"}
            if window_id == parasite and not provoked:
                provoked = True
                infected.add(parasite)
                spawn_pending = True
            elif window_id not in infected:
                live.remove(window_id)
            continue
        if kind == "spawn":
            listed = set(str(item) for item in event.get("echo_ids") or [])
            if not spawn_pending or str(event.get("parent_id") or "") != parasite or listed != echoes:
                return {"graded": True, "passed": False, "feedback": "parasite echo set was forged"}
            live.update(echoes)
            infected.update(echoes)
            spawn_pending = False
            continue
        if kind == "drag":
            if window_id not in live:
                return {"graded": True, "passed": False, "feedback": "dragged window was not live"}
            samples = event.get("samples")
            if not isinstance(samples, list) or not samples or len(samples) > 80:
                return {"graded": True, "passed": False, "feedback": "window drag samples are missing"}
            clean = []
            for point in samples:
                if not isinstance(point, list) or len(point) != 2 or not all(isinstance(value, int) for value in point):
                    return {"graded": True, "passed": False, "feedback": "window drag sample is malformed"}
                if clean and math.hypot(point[0] - clean[-1][0], point[1] - clean[-1][1]) > 120:
                    return {"graded": True, "passed": False, "feedback": "window teleported during containment"}
                clean.append(point)
            last_drag[window_id] = clean
            continue
        if kind == "contain":
            if not provoked or window_id not in infected or window_id not in live or window_id not in last_drag:
                return {"graded": True, "passed": False, "feedback": "containment occurred before an infected window was physically discovered"}
            source = popup_by_id.get(parasite, {}) if window_id in echoes else popup_by_id.get(window_id, {})
            end = last_drag[window_id][-1]
            cx, cy = end[0] + float(source.get("w") or 0) / 2, end[1] + float(source.get("h") or 0) / 2
            well = ground_truth.get("containment") or {}
            if not (float(well.get("x")) <= cx <= float(well.get("x")) + float(well.get("w")) and float(well.get("y")) <= cy <= float(well.get("y")) + float(well.get("h"))):
                return {"graded": True, "passed": False, "feedback": "infected window did not enter the visible containment well"}
            contained = window_id
            continue
        if kind == "purge":
            if purged or not contained or str(event.get("contained_id") or "") != contained:
                return {"graded": True, "passed": False, "feedback": "purge was not triggered by the contained parasite"}
            remaining = set(str(item) for item in event.get("remaining_before") or [])
            if remaining != live:
                return {"graded": True, "passed": False, "feedback": "purge inventory did not match the live desktop"}
            live.clear()
            purged = True
            continue
        return {"graded": True, "passed": False, "feedback": f"unknown containment event {kind}"}
    passed = provoked and not spawn_pending and bool(contained) and purged and not live
    return {"graded": True, "passed": passed, "feedback": f"parasite provoked={provoked}; infected contained={bool(contained)}; live windows={len(live)}"}

File: benchmarks/weird_captcha_gym/shared_runtime/app/mechanics/popup_exorcist.js
(() => {
  "use strict";
  let model = null;
  const esc = (value) => String(value == null ? "" : value).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const COPY = {
    update:["OPTIONAL UPDATE","A component may be available."],coupon:["BONUS WINDOW","Terms continue elsewhere."],cleaner:["DESKTOP CARE","Several pixels require attention."],forecast:["LOCAL FORECAST","A strong chance of more windows."],player:["MEDIA HELPER","Playback has not been requested."],survey:["ONE QUESTION","Closing is also an answer."],prize:["CLAIM PENDING","No prize was actually described."]
  };
  function record(kind, details={}){const event={sequence:model.events.length+1,kind,...details};model.events.push(event);return event;}
  function windowMarkup(item, infected=false){const copy=COPY[item.theme]||COPY.update;return `<article class="parasite-window theme-${esc(item.theme)} ${infected?"is-infected":""}" data-window-id="${esc(item.id)}" data-behavior="${esc(item.runtime_behavior||"echo")}" style="left:${item.x}px;top:${item.y}px;width:${item.w}px;height:${item.h}px;z-index:${item.z}"><header><i></i><span>${esc(item.title)}</span><button type="button" class="parasite-close" aria-label="Close">×</button></header><div class="parasite-body"><b>${esc(copy[0])}</b><p>${esc(copy[1])}</p><div class="fake-progress"><i></i></div></div></article>`;}
  function liveWindows(){return Array.from(document.querySelectorAll(".parasite-window:not(.is-dead)"));}
  function updateCount(){const node=document.querySelector(".parasite-count b");if(node)node.textContent=String(liveWindows().length);}
  function focus(node){node.style.zIndex=String(++model.topZ);record("focus",{window_id:node.dataset.windowId});}
  function activateContainment(){model.provoked=true;document.querySelector(".containment-well").dataset.active="true";document.querySelector(".popup-field").classList.add("is-contaminated");}
  async function failDesktop(){
    if(model.submitting)return;
    model.submitting=true;
    try{
      const response=await fetch("/result",{method:"POST",headers:{"content-type":"application/json"},body:JSON.stringify({mechanic_id:model.state.mechanic_id,challenge_id:model.state.challenge_id,events:model.events})});
      const outcome=await response.json();
      model.helpers.setReadout("FAIL","error");
      setTimeout(()=>outcome.state&&model.helpers.render(outcome.state),850);
    }catch(_error){model.submitting=false;model.helpers.setReadout("FAIL","error");}
  }
  function spawnEchoes(parent){const field=document.querySelector(".popup-field");const base={x:parseFloat(parent.style.left),y:parseFloat(parent.style.top),w:parent.offsetWidth,h:parent.offsetHeight,theme:parent.dataset.theme||"update",title:"DESKTOP MESSAGE"};model.state.echo_ids.forEach((id,index)=>{const item={...base,id,x:Math.max(8,Math.min(690-base.w,base.x+(index?74:-58))),y:Math.max(10,Math.min(365-base.h,base.y+54+index*22)),z:++model.topZ,runtime_behavior:"echo"};field.insertAdjacentHTML("beforeend",windowMarkup(item,true));const node=field.querySelector(`[data-window-id="${CSS.escape(id)}"]`);installWindow(node);});record("spawn",{parent_id:parent.dataset.windowId,echo_ids:[...model.state.echo_ids]});updateCount();}
  async function submit(containedId){
    if(model.submitting)return;
    model.submitting=true;
    record("purge",{contained_id:containedId,remaining_before:liveWindows().map(node=>node.dataset.windowId)});
    document.querySelector(".popup-field").classList.add("is-purging");
    liveWindows().forEach((node,index)=>setTimeout(()=>node.classList.add("is-dead"),index*75));
    setTimeout(updateCount,650);
    await new Promise(resolve=>setTimeout(resolve,700));
    try{
      const response=await fetch("/result",{method:"POST",headers:{"content-type":"application/json"},body:JSON.stringify({mechanic_id:model.state.mechanic_id,challenge_id:model.state.challenge_id,events:model.events})});
      const outcome=await response.json();
      if(outcome.passed===true){
        model.helpers.setReadout("PASS","passed");
        document.querySelector(".parasite-captcha").classList.add("is-passed");
      }else{
        model.helpers.setReadout("FAIL","error");
        setTimeout(()=>outcome.state&&model.helpers.render(outcome.state),850);
      }
    }catch(_error){
      model.submitting=false;
      model.helpers.setReadout("FAIL","error");
    }
  }
  function tryContain(node){if(!model.provoked||!node.classList.contains("is-infected"))return false;const well=document.querySelector(".containment-well").getBoundingClientRect();const rect=node.getBoundingClientRect();const cx=rect.left+rect.width/2,cy=rect.top+rect.height/2;if(cx<well.left||cx>well.right||cy<well.top||cy>well.bottom)return false;record("contain",{window_id:node.dataset.windowId});node.classList.add("is-contained");submit(node.dataset.windowId);return true;}
  function installWindow(node){if(!node)return;node.dataset.theme=node.className.match(/theme-([^ ]+)/)?.[1]||"update";node.addEventListener("pointerdown",()=>focus(node));const header=node.querySelector("header");header.addEventListener("pointerdown",event=>{if(event.target.closest("button"))return;event.preventDefault();focus(node);header.setPointerCapture(event.pointerId);const field=document.querySelector(".popup-field").getBoundingClientRect();const start=[event.clientX,event.clientY],origin=[parseFloat(node.style.left),parseFloat(node.style.top)];const samples=[];const move=moveEvent=>{const x=Math.max(0,Math.min(700-node.offsetWidth,origin[0]+moveEvent.clientX-start[0]));const y=Math.max(0,Math.min(390-node.offsetHeight,origin[1]+moveEvent.clientY-start[1]));node.style.left=`${x}px`;node.style.top=`${y}px`;samples.push([Math.round(x),Math.round(y)]);};const up=()=>{header.removeEventListener("pointermove",move);header.removeEventListener("pointerup",up);record("drag",{window_id:node.dataset.windowId,samples:samples.slice(-80)});tryContain(node);};header.addEventListener("pointermove",move);header.addEventListener("pointerup",up);});node.querySelector(".parasite-close").addEventListener("click",()=>{if(model.submitting)return;focus(node);const behavior=node.dataset.behavior;record("close",{window_id:node.dataset.windowId});if(behavior==="replicate"&&!model.provoked){node.classList.add("is-infected");activateContainment();spawnEchoes(node);return;}if(node.classList.contains("is-infected")){node.classList.remove("pulse");void node.offsetWidth;node.classList.add("pulse");model.strikes+=1;record("resist",{window_id:node.dataset.windowId,strike:model.strikes});if(model.strikes>=3)failDesktop();return;}node.classList.add("is-dead");updateCount();});}
  async function render(state,helpers){document.body.dataset.mechanic="popup-exorcist-v2";model={state,helpers,events:[],topZ:20,provoked:false,strikes:0,submitting:false};helpers.app.innerHTML=`<section class="parasite-captcha"><header class="parasite-head"><div><span>DESKTOP CONTAINMENT / LIVE FIELD</span><h1>${esc(state.prompt)}</h1></div><div class="parasite-count">WINDOWS <b>${state.popups.length}</b></div></header><section class="popup-field"><div class="field-wallpaper"><i></i><span>LOCAL DESKTOP</span></div><div class="containment-well" data-active="false" style="left:${state.containment.x}px;top:${state.containment.y}px;width:${state.containment.w}px;height:${state.containment.h}px"><i></i><b>⌁</b></div>${state.popups.map(item=>windowMarkup(item)).join("")}</section><footer class="parasite-foot"><div class="readout" data-status="idle"></div><span class="field-signal">CONTAINMENT FIELD: DORMANT</span></footer></section>`;document.querySelectorAll(".parasite-window").forEach(installWindow);const observer=new MutationObserver(()=>{const signal=document.querySelector(".field-signal");if(signal)signal.textContent=model.provoked?"CONTAINMENT FIELD: RESONANT":"CONTAINMENT FIELD: DORMANT";});observer.observe(document.querySelector(".containment-well"),{attributes:true});}
  window.WeirdCaptchaMechanics=window.WeirdCaptchaMechanics||{};window.WeirdCaptchaMechanics.popup_exorcist={render,rootSelector:".parasite-captcha"};
})();
